### PR TITLE
Implement methods 'unsqueeze()' and 'cat()'

### PR DESCRIPTION
**Pull request by litcoderr**
This revision includes:
- Implement methods 'unsequeeze()' and 'cat()'
  - `unsqueeze`: Expand a tensor's dimension
  - `cat`: Concatenate multiple tensors along a desired axis

### CODE REVIEW

File: src/tensor/mod.rs [hunk 1]
```diff
@@ -176,6 +176,73 @@ impl<T: traits::TensorTrait<T>> Tensor<T> {
...
+    pub fn unsqueeze(input: Tensor<T>, dim: usize) -> Tensor<T> {
+        let mut shape = input.shape.clone();
+        shape.insert(dim, 1);
+
+        Tensor::build(shape, input.components.clone())
+    }
+
+    pub fn cat(tensors: Vec<&Tensor<T>>, dim: usize) -> Tensor<T> {
+        let mut new_shape = tensors[0].shape.clone();
+        new_shape[dim] = tensors.len() as i32;
```
Comment (utilForever) [inline src/tensor/mod.rs:215]:
```suggestion
        let mut new_shapes = tensors[0].shape.clone();
        new_shapes[dim] = tensors.len() as i32;
```

File: src/tensor/mod.rs [hunk 2]
```diff
@@ -176,6 +176,73 @@ impl<T: traits::TensorTrait<T>> Tensor<T> {
...
+
+        let origin_shape_pi = Tensor::<T>::get_shape_pi(&tensors[0].shape);
+        let new_shape_pi = Tensor::<T>::get_shape_pi(&new_shape);
```
Comment (utilForever) [inline src/tensor/mod.rs:218]:
```suggestion
        let new_shape_pi = Tensor::<T>::get_shape_pi(&new_shapes);
```

File: src/tensor/mod.rs [hunk 3]
```diff
@@ -176,6 +176,73 @@ impl<T: traits::TensorTrait<T>> Tensor<T> {
...
+
+        let mut components: Vec<T> = vec![];
+
+        for tensor in &tensors {
+            for idx in 0..new_shape.len() {
+                if idx != dim && new_shape[idx] != tensor.shape[idx] {
+                    panic!("The tensors' shape should be the same except for the concatenation dimension");
+                }
+            }
+        }
```
Comment (utilForever) [inline src/tensor/mod.rs]:
```suggestion
        tensors.iter().zip(new_shapes.iter().enumerate()).for_each(|(tensor, (idx, new_shape))| {
            if idx != dim && tensor.shape[idx] != *new_shape {
                panic!("The tensors' shape should be the same except for the concatenation dimension");
            }
        });
```

File: src/tensor/mod.rs [hunk 4]
```diff
@@ -176,6 +176,73 @@ impl<T: traits::TensorTrait<T>> Tensor<T> {
...
+
+        // TODO build components
+        let mut new_size = 1;
+        for s in &new_shape {
+            new_size *= s;
+        }
```
Comment (utilForever) [inline src/tensor/mod.rs]:
```suggestion
        let new_size = new_shapes.iter().fold(1, |acc, x| acc * x);
```

File: src/tensor/mod.rs [hunk 5]
```diff
@@ -176,6 +176,73 @@ impl<T: traits::TensorTrait<T>> Tensor<T> {
...
+                if shape_idx == dim {
+                    tensor_idx = val as usize;
+                } else {
+                    origin_idx = origin_idx + (val * origin_shape_pi[shape_idx]) as usize;
+                }
+
+                new_idx = new_idx - val * pi;
+            }
+
+            let origin_component= tensors[tensor_idx].components[origin_idx].clone();
```
Comment (utilForever) [inline src/tensor/mod.rs]:
```suggestion
            let origin_component = tensors[tensor_idx].components[origin_idx].clone();
```

File: src/tensor/mod.rs [hunk 6]
```diff
@@ -176,6 +176,73 @@ impl<T: traits::TensorTrait<T>> Tensor<T> {
...
+            components.push(origin_component);
+        }
+
+        Tensor::build(new_shape, components)
```
Comment (utilForever) [inline src/tensor/mod.rs:250]:
```suggestion
        Tensor::build(new_shapes, components)
```

File: src/tensor/mod.rs [hunk 7]
```diff
@@ -176,6 +176,73 @@ impl<T: traits::TensorTrait<T>> Tensor<T> {
...
+    }
+
+    // utility methods
+
+    fn get_shape_pi(shape: &Vec<i32>) -> Vec<i32> {
+        let mut shape_pi: Vec<i32> = vec![1];
+
+        for idx in (0..shape.len()-1).rev() {
+            let pi = shape[idx+1] * shape_pi[0];
```
Comment (utilForever) [inline src/tensor/mod.rs]:
```suggestion
        for idx in (0..shape.len() - 1).rev() {
            let pi = shape[idx + 1] * shape_pi[0];
```

**Review by utilForever — CHANGES_REQUESTED** [review]
You should run `cargo fmt` and `cargo clippy`.

**Review by utilForever — APPROVED** [review]
LGTM now. Thanks for your work! 🦖

**Comment by litcoderr** [comment]
No problemo

**Pull request merged**